### PR TITLE
Make BOXContentSDK extension-safe : introducing an UIApplication category that will avoid compiling forbidden extensions APIs.

### PR DESCRIPTION
**Pull request by crousselle**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @crousselle has signed the CLA. Thanks for the pull request!

File: BoxContentSDK/BoxContentSDK/Categories/UIApplication+ExtensionSafeCompilation.m [hunk 1]
```diff
@@ -0,0 +1,47 @@
+//
+//  UIApplication+ExtensionSafeCompilation.m
+//  app
```
Comment (scottliubox) [inline BoxContentSDK/BoxContentSDK/Categories/UIApplication+ExtensionSafeCompilation.m]: probably delete "app" here. Same for .h

File: BoxContentSDK/BoxContentSDK/Categories/UIApplication+ExtensionSafeCompilation.m [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    return nil;
+#else
+    // If we are compiling from the main app, return normal sharedApplication.
+    return [UIApplication sharedApplication];
+#endif
+}
+
+- (BOOL)box_canOpenURL:(NSURL *)url
+{
+#ifdef EXTENSION
```
Comment (scottliubox) [inline BoxContentSDK/BoxContentSDK/Categories/UIApplication+ExtensionSafeCompilation.m]: maybe re-name this to something more explicit? Maybe CompileForExtension

Comment (crousselle) [inline BoxContentSDK/BoxContentSDK/Categories/UIApplication+ExtensionSafeCompilation.m]: this should be to specify that the current target is an extension. What about TARGET_IS_EXTENSION ?

Comment (scottliubox) [inline BoxContentSDK/BoxContentSDK/Categories/UIApplication+ExtensionSafeCompilation.m]: That sounds good, let's go with that.

File: BoxContentSDK/BoxContentSDK/Categories/UIApplication+ExtensionSafeCompilation.m [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#ifdef TARGET_IS_EXTENSION
+    // If we are compiling from an extension, return nil here since UIApplication isn't extension-safe.
+    return nil;
+#else
+    // If we are compiling from a non-extension target, use the regular sharedApplication.
+    return [UIApplication sharedApplication];
+#endif
+}
+
+- (BOOL)box_canOpenURL:(NSURL *)url
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Categories/UIApplication+ExtensionSafeCompilation.m]: why do we need these? If you call box_sharedApplication you're set. Is there a reason we can't use box_SharedApplication in some cases?

File: BoxContentSDK/BoxContentSDK/OAuth2/BOXAuthorizationViewController.m [hunk 4]
```diff
@@ -355,7 +356,7 @@ - (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)
 
     // Mailto URLs can be encountered if there is a hyperlink for sending an email.
     if ([[[[request URL] scheme] lowercaseString] isEqualToString:@"mailto"]) {
-        [[UIApplication sharedApplication] openURL:[request URL]];
+        [[UIApplication box_sharedApplication] box_openURL:[request URL]];
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/OAuth2/BOXAuthorizationViewController.m:359]: couldn't this just call the regular openURL method? Or will that not compile for extensions??

Comment (crousselle) [inline BoxContentSDK/BoxContentSDK/OAuth2/BOXAuthorizationViewController.m:359]: @scottliubox is there any resaons behind those implementations (this is the category that was in the app)

Comment (crousselle) [inline BoxContentSDK/BoxContentSDK/OAuth2/BOXAuthorizationViewController.m:359]: @boxtcarpel somehow Xcode throws an error if we do not use our implementations. Hence the category methods.

Comment (scottliubox) [inline BoxContentSDK/BoxContentSDK/OAuth2/BOXAuthorizationViewController.m:359]: XCode actually throws an error for using openURL itself, it specifically mentions that method not being extension safe and doesn't let you compile.

**Comment by scottliubox** [comment]
+2

**Pull request merged**